File: services/control-api/src/cloudctl_api/builtin_recipes.py
"""Frozen APK-local Recipe packages referenced by CommandV1.

These are not a published catalog. P14 still owns signed rollout. Claim and
Companion both pin the same versionId+sha256 so a phone can load the matching
local graph without turning signature verification off.
"""

from __future__ import annotations

import hashlib
import json
from typing import Any

_ENGINE_MIN_VERSION = 1


def _canonical_recipe_bytes(package: dict[str, Any]) -> bytes:
    body = {
        "apiVersion": package["apiVersion"],
        "kind": package["kind"],
        "manifest": {key: value for key, value in package["manifest"].items() if key != "hash"},
        "graph": package["graph"],
    }
    return json.dumps(body, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode()


def _freeze(package: dict[str, Any]) -> dict[str, Any]:
    frozen = json.loads(json.dumps(package))
    frozen["manifest"]["hash"] = hashlib.sha256(_canonical_recipe_bytes(frozen)).hexdigest()
    return frozen


def _package(
    *,
    recipe_id: str,
    platform: str,
    app: str,
    command_type: str,
    states: list[dict[str, Any]],
    start_state_id: str,
    commit_action_id: str | None = None,
) -> dict[str, Any]:
    graph: dict[str, Any] = {
        "startStateId": start_state_id,
        "maxIterations": 8,
        "maxDurationMs": 30_000,
        "states": states,
    }
    if commit_action_id:
        graph["commitActionId"] = commit_action_id
    return _freeze(
        {
            "apiVersion": "cloudctl.recipe/v1",
            "kind": "LocalRecipePackage",
            "manifest": {
                "id": recipe_id,
                "version": "1.0.0",
                "hash": "0" * 64,
                "signingKeyId": "builtin-phase1",
                "minEngineVersion": _ENGINE_MIN_VERSION,
                "platform": platform,
                "app": app,
                "commandTypes": [command_type],
            },
            "graph": graph,
            "signature": {
                "algorithm": "Ed25519",
                "keyId": "builtin-phase1",
                "digest": "hash-pinned-builtin",
            },
        }
    )


# B05 (K05 platform-recipe/v1 §4 open-only): the minimal xianyu publish graph.
# Flow mirrors the frozen dispatch-xianyu steps family (tapsPublish=false) and
# stops at the confirm point — the checkpoint state is terminal WAITING_USER and
# the graph contains no publish/submit action of any kind (xianyu_publish_button
# is never referenced; real submission stays behind Q02 pre-protection +
# per-object authorization). Media selection is expressed by the single "media"
# state: the companion engine expands it to the ordered tile taps
# xianyu_gallery_select_1..N (tile 0 is the camera shutter and is never
# referenced; mediaDelivery item order == download/export order == tile order).
# Task values are bound at execution time from CommandV1 parameters
# (xianyu_description <- listingBody, xianyu_price <- price); the graph bytes
# stay parameter-free so the hash is stable.
_XIANYU_PUBLISH_OPEN_ONLY_STATES: list[dict[str, Any]] = [
    {
        "stateId": "wait-home",
        "action": "wait",
        "locatorRef": "xianyu_home_sell",
        "onSuccess": "open-sell",
        "onFailure": "FAILED",
    },
    {
        "stateId": "open-sell",
        "action": "tap",
        "locatorRef": "xianyu_home_sell",
        "postcondition": "xianyu_publish_entry",
        "onSuccess": "open-publish",
        "onFailure": "FAILED",
    },
    {
        "stateId": "open-publish",
        "action": "tap",
        "locatorRef": "xianyu_publish_entry",
        "postcondition": "xianyu_publish_page",
        "onSuccess": "await-form",
        "onFailure": "FAILED",
    },
    {
        "stateId": "await-form",
        "action": "wait",
        "locatorRef": "xianyu_publish_page",
        "onSuccess": "select-media",
        "onFailure": "FAILED",
    },
    {
        "stateId": "select-media",
        "action": "media",
        "locatorRef": "xianyu_add_image",
        "onSuccess": "fill-description",
        "onFailure": "FAILED",
    },
    {
        "stateId": "fill-description",
        "action": "input",
        "locatorRef": "xianyu_description",
        "valueRef": "listingBody",
        "onSuccess": "confirm-description",
        "onFailure": "FAILED",
    },
    {
        "stateId": "confirm-description",
        "action": "tap",
        "locatorRef": "xianyu_composer_done",
        "onSuccess": "await-price",
        "onFailure": "FAILED",
    },
    {
        "stateId": "await-price",
        "action": "wait",
        "locatorRef": "xianyu_price",
        "onSuccess": "fill-price",
        "onFailure": "FAILED",
    },
    {
        "stateId": "fill-price",
        "action": "input",
        "locatorRef": "xianyu_price",
        "valueRef": "price",
        "onSuccess": "capture-confirm-point",
        "onFailure": "FAILED",
    },
    {
        "stateId": "capture-confirm-point",
        "action": "extract",
        "onSuccess": "await-confirm",
        "onFailure": "FAILED",
    },
    {
        "stateId": "await-confirm",
        "action": "checkpoint",
        "onSuccess": "WAITING_USER",
        "onPause": "WAITING_USER",
        "terminal": True,
    },
]

_XIANYU_PUBLISH_OPEN_ONLY = _package(
    recipe_id="recipe-xianyu-publish-2",
    platform="xianyu",
    app="com.taobao.idlefish",
    command_type="xianyu.publish_listing.v1",
    start_state_id="wait-home",
    states=_XIANYU_PUBLISH_OPEN_ONLY_STATES,
)
_XIANYU_PUBLISH_OPEN_ONLY["graph"]["maxIterations"] = 40
_XIANYU_PUBLISH_OPEN_ONLY["graph"]["maxDurationMs"] = 600_000
_XIANYU_PUBLISH_OPEN_ONLY = _freeze(_XIANYU_PUBLISH_OPEN_ONLY)

# Byte-frozen v1 open-only stub (hash a2331b80…). Tasks claimed before the v2
# graph pinned this exact package; their pins resolve on the phone from the
# APK-local literal, and the server keeps the frozen bytes here for hash
# audit/verification. New claims of xianyu.publish_listing.v1 resolve the v2
# graph above (family-A first-claim pin, K05 §3).
_LEGACY_XIANYU_PUBLISH_OPEN_ONLY = _package(
    recipe_id="recipe-xianyu-publish-1",
    platform="xianyu",
    app="com.taobao.idlefish",
    command_type="xianyu.publish_listing.v1",
    start_state_id="open-only",
    states=[
        {
            "stateId": "open-only",
            "action": "checkpoint",
            "onSuccess": "WAITING_USER",
            "onPause": "WAITING_USER",
            "terminal": True,
        }
    ],
)

LEGACY_BUILTIN_PACKAGES: dict[str, dict[str, Any]] = {
    "recipe-xianyu-publish-1": _LEGACY_XIANYU_PUBLISH_OPEN_ONLY,
}

BUILTIN_RECIPES: dict[str, dict[str, Any]] = {
    "device.probe_capabilities.v1": _package(
        recipe_id="recipe-device-probe-1",
        platform="companion",
        app="com.company.cloudctl.companion",
        command_type="device.probe_capabilities.v1",
        start_state_id="probe",
        states=[
            {
                "stateId": "probe",
                "action": "log",
                "onSuccess": "SUCCEEDED",
                "terminal": True,
            }
        ],
    ),
    "xianyu.collect_orders.v1": _package(
        recipe_id="recipe-xianyu-collect-1",
        platform="xianyu",
        app="com.taobao.idlefish",
        command_type="xianyu.collect_orders.v1",
        start_state_id="collect",
        states=[
            {
                "stateId": "collect",
                "action": "log",
                "onSuccess": "SUCCEEDED",
                "terminal": True,
            }
        ],
    ),
    "xianyu.publish_listing.v1": _XIANYU_PUBLISH_OPEN_ONLY,
    "xiaohongshu.publish_note.v1": _package(
        recipe_id="recipe-xhs-note-1",
        platform="xiaohongshu",
        app="com.xingin.xhs",
        command_type="xiaohongshu.publish_note.v1",
        start_state_id="open-only",
        states=[
            {
                "stateId": "open-only",
                "action": "checkpoint",
                "onSuccess": "WAITING_USER",
                "onPause": "WAITING_USER",
                "terminal": True,
            }
        ],
    ),
}

OPEN_ONLY_COMMAND_TYPES = frozenset(
    {
        "xianyu.publish_listing.v1",
        "xiaohongshu.publish_note.v1",
    }
)


def builtin_recipe_ref(command_type: str) -> dict[str, Any]:
    package = BUILTIN_RECIPES[command_type]
    return {
        "versionId": package["manifest"]["id"],
        "sha256": package["manifest"]["hash"],
        "engineMinVersion": package["manifest"]["minEngineVersion"],
    }


def builtin_recipe_package(command_type: str) -> dict[str, Any]:
    return json.loads(json.dumps(BUILTIN_RECIPES[command_type]))

File: mobile/companion/app/src/main/java/com/company/cloudctl/companion/automation/BuiltinRecipes.kt
package com.company.cloudctl.companion.automation

import org.json.JSONObject

object BuiltinRecipes {
    val OPEN_ONLY_COMMAND_TYPES = setOf(
        "xianyu.publish_listing.v1",
        "xiaohongshu.publish_note.v1",
    )

    fun packageJson(versionId: String): String? = PACKAGES[versionId]

    fun jsonFor(command: CommandV1): String {
        val encoded = packageJson(command.recipeVersionId)
            ?: error(CommandV1Parser.UNSUPPORTED_RECIPE)
        val root = JSONObject(encoded)
        val manifest = root.getJSONObject("manifest")
        require(manifest.getString("hash") == command.recipeSha256) { "recipe hash mismatch" }
        require(manifest.getString("app") == command.targetPackage) { "recipe app does not match command" }
        require(manifest.getJSONArray("commandTypes").getString(0) == command.commandType) {
            CommandV1Parser.UNSUPPORTED_RECIPE
        }
        return encoded
    }

    private val PACKAGES = mapOf(
        "recipe-device-probe-1" to """{"apiVersion":"cloudctl.recipe/v1","kind":"LocalRecipePackage","manifest":{"id":"recipe-device-probe-1","version":"1.0.0","hash":"901f795b2512891cd5ee08162626d0ad6a1c92ad9d7b922516c626f9d6f155b4","signingKeyId":"builtin-phase1","minEngineVersion":1,"platform":"companion","app":"com.company.cloudctl.companion","commandTypes":["device.probe_capabilities.v1"]},"graph":{"startStateId":"probe","maxIterations":8,"maxDurationMs":30000,"states":[{"stateId":"probe","action":"log","onSuccess":"SUCCEEDED","terminal":true}]},"signature":{"algorithm":"Ed25519","keyId":"builtin-phase1","digest":"hash-pinned-builtin"}}""",
        "recipe-xianyu-collect-1" to """{"apiVersion":"cloudctl.recipe/v1","kind":"LocalRecipePackage","manifest":{"id":"recipe-xianyu-collect-1","version":"1.0.0","hash":"d567f2b8ea3f338af9360c5befc7124770ae92d17e932cb7b03cd6e3f2d4b4b3","signingKeyId":"builtin-phase1","minEngineVersion":1,"platform":"xianyu","app":"com.taobao.idlefish","commandTypes":["xianyu.collect_orders.v1"]},"graph":{"startStateId":"collect","maxIterations":8,"maxDurationMs":30000,"states":[{"stateId":"collect","action":"log","onSuccess":"SUCCEEDED","terminal":true}]},"signature":{"algorithm":"Ed25519","keyId":"builtin-phase1","digest":"hash-pinned-builtin"}}""",
        "recipe-xianyu-publish-1" to """{"apiVersion":"cloudctl.recipe/v1","kind":"LocalRecipePackage","manifest":{"id":"recipe-xianyu-publish-1","version":"1.0.0","hash":"a2331b8076dd28cd34f574f359b7e549ebef1f0ebd060bff2fad0ab410a84c0f","signingKeyId":"builtin-phase1","minEngineVersion":1,"platform":"xianyu","app":"com.taobao.idlefish","commandTypes":["xianyu.publish_listing.v1"]},"graph":{"startStateId":"open-only","maxIterations":8,"maxDurationMs":30000,"states":[{"stateId":"open-only","action":"checkpoint","onSuccess":"WAITING_USER","onPause":"WAITING_USER","terminal":true}]},"signature":{"algorithm":"Ed25519","keyId":"builtin-phase1","digest":"hash-pinned-builtin"}}""",
        "recipe-xianyu-publish-2" to """{"apiVersion":"cloudctl.recipe/v1","graph":{"maxDurationMs":600000,"maxIterations":40,"startStateId":"wait-home","states":[{"action":"wait","locatorRef":"xianyu_home_sell","onFailure":"FAILED","onSuccess":"open-sell","stateId":"wait-home"},{"action":"tap","locatorRef":"xianyu_home_sell","onFailure":"FAILED","onSuccess":"open-publish","postcondition":"xianyu_publish_entry","stateId":"open-sell"},{"action":"tap","locatorRef":"xianyu_publish_entry","onFailure":"FAILED","onSuccess":"await-form","postcondition":"xianyu_publish_page","stateId":"open-publish"},{"action":"wait","locatorRef":"xianyu_publish_page","onFailure":"FAILED","onSuccess":"select-media","stateId":"await-form"},{"action":"media","locatorRef":"xianyu_add_image","onFailure":"FAILED","onSuccess":"wait-edit-page","stateId":"select-media"},{"action":"wait","locatorRef":"xianyu_crop_done","onFailure":"FAILED","onSuccess":"confirm-crop","stateId":"wait-edit-page"},{"action":"tap","locatorRef":"xianyu_crop_done","onFailure":"FAILED","onSuccess":"await-form-back","stateId":"confirm-crop"},{"action":"wait","locatorRef":"xianyu_publish_page","onFailure":"FAILED","onSuccess":"fill-description","stateId":"await-form-back"},{"action":"input","locatorRef":"xianyu_description","onFailure":"FAILED","onSuccess":"capture-confirm-point","stateId":"fill-description","valueRef":"listingBody"},{"action":"extract","onFailure":"FAILED","onSuccess":"await-confirm","stateId":"capture-confirm-point"},{"action":"checkpoint","onPause":"WAITING_USER","onSuccess":"WAITING_USER","stateId":"await-confirm","terminal":true}]},"kind":"LocalRecipePackage","manifest":{"app":"com.taobao.idlefish","commandTypes":["xianyu.publish_listing.v1"],"hash":"f6adebdca3575cce16b64b62b6e27730de77868d64cceff3e83bee2223f62372","id":"recipe-xianyu-publish-2","minEngineVersion":1,"platform":"xianyu","signingKeyId":"builtin-phase1","version":"1.1.0"},"signature":{"algorithm":"Ed25519","digest":"hash-pinned-builtin","keyId":"builtin-phase1"}}""",
        "recipe-xhs-note-1" to """{"apiVersion":"cloudctl.recipe/v1","kind":"LocalRecipePackage","manifest":{"id":"recipe-xhs-note-1","version":"1.0.0","hash":"b1bdc6f334d2cea210c544183ff999244dfe4fe052b621d6aff6ddb9f495fda4","signingKeyId":"builtin-phase1","minEngineVersion":1,"platform":"xiaohongshu","app":"com.xingin.xhs","commandTypes":["xiaohongshu.publish_note.v1"]},"graph":{"startStateId":"open-only","maxIterations":8,"maxDurationMs":30000,"states":[{"stateId":"open-only","action":"checkpoint","onSuccess":"WAITING_USER","onPause":"WAITING_USER","terminal":true}]},"signature":{"algorithm":"Ed25519","keyId":"builtin-phase1","digest":"hash-pinned-builtin"}}""",
    )
}
